### PR TITLE
fix the scope of optimizer

### PR DESCRIPTION
**Pull request by soskek**
The target of an optimizer should contain `clf_head` (new task-specific output matrix) in addition to `model` (Transformer encoder).
The code might fail to do that, right?

### CODE REVIEW

**Comment by thomwolf** [comment]
You are right! Thanks a lot!!
But shouldn't we also add the parameters of the language modeling head `LMHead` so that the gradients on the embeddings (output is tied to the input) are computed well?
I will try to do a quick training with and without to check we are better with them.

**Comment by soskek** [comment]
That's also a good point.
Though I'm not so familiar with pytorch, it may be also possible to replace
```
        self.decoder = nn.Linear(cfg.n_embd, model.vocab, bias=False)
        self.decoder.weight = model.embed.weight # Tied weights
```
with
```
        self.decoder = lambda x: F.linear(x, model.embed.weight)
```

BTW, I ported your great code to one using Chainer. Thanks!
https://github.com/soskek/chainer-openai-transformer-lm
(Through the poring, I sent the PRs and found this method.)

**Comment by thomwolf** [comment]
Yes I think it should work and it's more elegant indeed.

Thanks for sharing your port, very nice to see it in Chainer!
